pkg/oc/util/tokencmd/negotiator_sspi.go

```diff
@@ -0,0 +1,188 @@
+		sspi.ISC_REQ_INTEGRITY |
+		sspi.ISC_REQ_MUTUAL_AUTH |
+		sspi.ISC_REQ_REPLAY_DETECT |
+		sspi.ISC_REQ_SEQUENCE_DETECT
+
+	// separator used in fully qualified user name format
+	domainSeparator = `\`
+
+	// max lengths for various fields, see sspiNegotiator.principalName
+	maxUsername = 256
+	maxPassword = 256
+	maxDomain   = 15
+)
+
+func SSPIEnabled() bool {
+	return true
+}
+
+// sspiNegotiator handles negotiate flows on windows via SSPI
+// It expects sspiNegotiator.InitSecContext to be called until sspiNegotiator.IsComplete returns true
+type sspiNegotiator struct {
+	// optional DOMAIN\Username and password
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/aa374714(v=vs.85).aspx
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/aa380131(v=vs.85).aspx
+	// pAuthData [in]: If credentials are supplied, they are passed via a pointer to a sspi.SEC_WINNT_AUTH_IDENTITY
+	// structure that includes those credentials.
+	// When using the Negotiate package, the maximum character lengths for user name, password, and domain are
+	// 256, 256, and 15, respectively.
+	principalName string
+	password      string
+
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/ms721572(v=vs.85).aspx#_security_credentials_gly
+	// phCredential [in, optional]: A handle to the credentials returned by AcquireCredentialsHandle (Negotiate).
+	// This handle is used to build the security context.  sspi.SECPKG_CRED_OUTBOUND is used to request OUTBOUND credentials.
+	cred *sspi.Credentials
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/ms721625(v=vs.85).aspx#_security_security_context_gly
+	// Manages all steps of the Negotiate negotiation.
+	ctx *negotiate.ClientContext
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/aa375509(v=vs.85).aspx
+	// fContextReq [in]: Bit flags that indicate requests for the context.
+	flags uint32
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/aa375509(v=vs.85).aspx
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/aa374764(v=vs.85).aspx
+	// Set to true once InitializeSecurityContext or CompleteAuthToken return sspi.SEC_E_OK
+	complete bool
+}
+
+func NewSSPINegotiator(principalName, password string) Negotiator {
+	return &sspiNegotiator{principalName: principalName, password: password, flags: flags}
+}
+
+func (s *sspiNegotiator) Load() error {
+	glog.V(5).Info("Attempt to load SSPI")
+	// do nothing since SSPI uses lazy DLL loading
+	return nil
+}
+
+func (s *sspiNegotiator) InitSecContext(requestURL string, challengeToken []byte) ([]byte, error) {
+	defer runtime.HandleCrash()
+	if s.cred == nil || s.ctx == nil {
+		glog.V(5).Infof("Start SSPI flow: %s", requestURL)
+
+		cred, err := s.getUserCredentials()
+		if err != nil {
+			glog.V(5).Infof("getUserCredentials returned error: %v", err)
+			return nil, err
+		}
+		s.cred = cred
+		glog.V(5).Info("getUserCredentials successful")
+
+		serviceName, err := getServiceName('/', requestURL)
```

Comment:
I am thinking getServiceName should be implented in each module and not a helper function, given the form is different.
For example, although not used currently, Microsoft can compose tickets where the "port" is also use to compose the names, while this is normally not possible with GSSAPI, so the name should be built individually by negotiator implementation

Reply:
How would the SSPI code know it was support to use the port or not?

Reply:
it wouldn't, it is kind of service specific, we'd need a config option I guess.
But the point I was making is that rules can be different, so perhaps a custom function is more flexible in the long run, it's just a nit, not a big deal